TOOLS/lua/autocrop.lua

```diff
@@ -1,114 +1,341 @@
+    the value is larger than the remaining playback time.
+
+detect_limit_num: number[0-255] &
+detect_limit_den: number[0-255] - Black threshold for cropdetect.
+    The final result is detect_limit_num / detect_limit_den.
+    Smaller values will generally result in less cropping.
+    See limit of https://ffmpeg.org/ffmpeg-filters.html#cropdetect
+
+
 
-script_name = mp.get_script_name()
-cropdetect_label = string.format("%s-cropdetect", script_name)
-crop_label = string.format("%s-crop", script_name)
+detect_round: number[2^n] -  The value which the width/height
+    should be divisible by. Smaller values ​​have better detection
+    accuracy. If you have problems with other filters,
+    you can try to set it to 4 or 16.
+    See round of https://ffmpeg.org/ffmpeg-filters.html#cropdetect
 
--- number of seconds to gather cropdetect data
-detect_seconds = tonumber(mp.get_opt(string.format("%s.detect_seconds", script_name)))
-if not detect_seconds then
+detect_min_ratio: number[0.0-1.0] - The ratio of the minimum clip
+    size to the original. If the picture is over cropped or under
+    cropped, try adjusting this value.
+
+detect_seconds: seconds - How long to gather cropdetect data.
+    Increasing this may be desirable to allow cropdetect more
+    time to collect data.
+--]]
+
+require "mp.msg"
+require 'mp.options'
+
+local options = {
+    auto = true,
+    auto_delay = 4,
+    detect_limit_num = 24,
+    detect_limit_den = 255,
+    detect_round = 2,
+    detect_min_ratio = 0.5,
     detect_seconds = 1
-end
+}
+read_options(options)
+
+local label_prefix = mp.get_script_name()
+local labels = {
+    crop = string.format("%s-crop", label_prefix),
+    cropdetect = string.format("%s-cropdetect", label_prefix)
+}
 
-function del_filter_if_present(label)
-    -- necessary because mp.command('vf del @label:filter') raises an
-    -- error if the filter doesn't exist
-    local vfs = mp.get_property_native("vf")
-    for i,vf in pairs(vfs) do
-        if vf["label"] == label then
-            table.remove(vfs, i)
-            mp.set_property_native("vf", vfs)
+timers = {
+    auto_delay = nil,
+    detect_crop = nil
+}
+
+function is_filter_present(label)
+    local filters = mp.get_property_native("vf")
+    for index, filter in pairs(filters) do
+        if filter["label"] == label then
             return true
         end
     end
     return false
 end
 
-function autocrop_start()
-    -- exit if cropdetection is already in progress
-    if timer then
-        mp.msg.warn("already cropdetecting!")
-        return
+function is_time_enough(seconds)
+
+    -- Plus 1 second for deviation.
+    local time_needed = seconds + 1
+
+    -- For still images
+    if typeof_playing() == "image" then
+        return true
+    end
+
+    return time_needed < mp.get_property_native("playtime-remaining")
+end
+
+-- The result of this function is not accurate
+-- Audio will return "unknown" and GIF will return "video"
+function typeof_playing()
```

Comment:
I think this is more information than we need. On top of that it doesn't really identify video vs. images because it will classify a video starting near the end or a short video as an image. I should only be necessary to identify album art. The check for sufficient playback time to run cropdetect will disallow images by themselves as a side effect.

Reply:
I try to simplify it like this
```
function is_cropable()
    local vid = mp.get_property_native("vid")
    local is_album = vid and mp.get_property_native(
        string.format("track-list/%d/albumart", vid)
    ) or false

    return vid and not is_album
end
```
still image will warn that there is not enough time, this is confusing.
frame count is also unreliable, any ideas?

Reply:
Well, the problem with the still image *is* that there's not enough time.